File: frontend/apps/ios/Pantopus/Features/Homes/Bills/AddBillWizardViewModel.swift
//
//  AddBillWizardViewModel.swift
//  Pantopus
//
//  3-step Wizard for `POST /api/homes/:id/bills` (route
//  `backend/routes/home.js:4539`).
//
//   1. Payee + amount + due date
//   2. Schedule (one-time / recurring monthly) — stored in `details.schedule`
//   3. Review & confirm — POSTs the bill
//
//  Backend caveat: there is no POST/PATCH for `:billId/splits` today, so
//  the design's "Split between household members" step is not modelled
//  here. That decision is tracked in the parity audit and surfaces in
//  the wizard as a note on the review step.
//

import Foundation
import Observation

/// Step identifiers for the Add Bill wizard.
public enum AddBillStep: String, Sendable, CaseIterable {
    case details
    case schedule
    case review
    case success
}

/// Recurrence options exposed on step 2.
public enum AddBillSchedule: String, Sendable, CaseIterable, Identifiable {
    case oneTime
    case monthly
    case quarterly
    case yearly

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .oneTime: "One-time"
        case .monthly: "Recurring monthly"
        case .quarterly: "Recurring quarterly"
        case .yearly: "Recurring yearly"
        }
    }

    public var detailsKey: String {
        switch self {
        case .oneTime: "one_time"
        case .monthly: "monthly"
        case .quarterly: "quarterly"
        case .yearly: "yearly"
        }
    }
}

/// Outbound events the wizard view must react to.
public enum AddBillOutboundEvent: Sendable, Equatable {
    case dismiss
    case created(billId: String)
}

@Observable
@MainActor
final class AddBillWizardViewModel: WizardModel {
    // Step 1 fields
    var payee: String = ""
    var amount: String = ""
    var dueDate: Date?

    // Step 2 fields
    var schedule: AddBillSchedule = .oneTime

    // Lifecycle
    private(set) var currentStep: AddBillStep = .details
    private(set) var isSubmitting: Bool = false
    private(set) var submitError: String?
    private(set) var createdBillId: String?
    var pendingEvent: AddBillOutboundEvent?

    private let homeId: String
    private let api: APIClient

    init(homeId: String, api: APIClient = .shared) {
        self.homeId = homeId
        self.api = api
    }

    // MARK: - WizardModel

    var chrome: WizardChrome {
        switch currentStep {
        case .details:
            WizardChrome(
                title: "Add a bill",
                progressLabel: .stepOf(current: 1, total: 3),
                progressFraction: 1.0 / 3.0,
                leading: .close,
                primaryCTALabel: "Next",
                primaryCTAEnabled: detailsValid,
                isSubmitting: false,
                dirty: isDirty,
                showsProgressBar: true
            )
        case .schedule:
            WizardChrome(
                title: "Schedule",
                progressLabel: .stepOf(current: 2, total: 3),
                progressFraction: 2.0 / 3.0,
                leading: .back,
                primaryCTALabel: "Next",
                primaryCTAEnabled: true,
                isSubmitting: false,
                dirty: isDirty,
                showsProgressBar: true
            )
        case .review:
            WizardChrome(
                title: "Review",
                progressLabel: .stepOf(current: 3, total: 3),
                progressFraction: 3.0 / 3.0,
                leading: .back,
                primaryCTALabel: "Add bill",
                primaryCTAEnabled: !isSubmitting,
                isSubmitting: isSubmitting,
                dirty: isDirty,
                showsProgressBar: true
            )
        case .success:
            WizardChrome(
                title: "Bill added",
                progressLabel: .hidden,
                progressFraction: nil,
                leading: .close,
                primaryCTALabel: "Done",
                primaryCTAEnabled: true,
                secondaryCTA: nil,
                isSubmitting: false,
                dirty: false,
                showsProgressBar: false
            )
        }
    }

    func leadingTapped() {
        switch currentStep {
        case .details:
            pendingEvent = .dismiss
        case .schedule:
            currentStep = .details
        case .review:
            currentStep = .schedule
        case .success:
            pendingEvent = .dismiss
        }
    }

    func discardConfirmed() {
        pendingEvent = .dismiss
    }

    func primaryTapped() {
        switch currentStep {
        case .details:
            currentStep = .schedule
            Analytics.track(.screenAddBillWizardStepViewed(stepNumber: 2, stepName: "schedule"))
        case .schedule:
            currentStep = .review
            Analytics.track(.screenAddBillWizardStepViewed(stepNumber: 3, stepName: "review"))
        case .review:
            Task { await submit() }
        case .success:
            if let id = createdBillId {
                pendingEvent = .created(billId: id)
            } else {
                pendingEvent = .dismiss
            }
        }
    }

    // MARK: - Validation

    var detailsValid: Bool {
        !payee.trimmingCharacters(in: .whitespaces).isEmpty
            && parsedAmount() != nil
    }

    var isDirty: Bool {
        !payee.trimmingCharacters(in: .whitespaces).isEmpty
            || !amount.trimmingCharacters(in: .whitespaces).isEmpty
            || dueDate != nil
            || schedule != .oneTime
    }

    func parsedAmount() -> Decimal? {
        let trimmed = amount.trimmingCharacters(in: .whitespaces)
        guard !trimmed.isEmpty,
              let value = Decimal(string: trimmed),
              value > 0
        else {
            return nil
        }
        return value
    }

    // MARK: - Submit

    func submit() async {
        guard let amountValue = parsedAmount(), !isSubmitting else { return }
        if !NetworkMonitor.shared.isOnline {
            submitError = "You're offline. Try again when you're back online."
            return
        }
        isSubmitting = true
        submitError = nil
        defer { isSubmitting = false }

        var details: [String: String] = ["schedule": schedule.detailsKey]
        if schedule != .oneTime {
            details["frequency"] = schedule.detailsKey
        }
        let request = CreateBillRequest(
            billType: "other",
            providerName: payee.trimmingCharacters(in: .whitespaces),
            amount: amountValue,
            dueDate: dueDate.map { AddBillWizardViewModel.formatISODate($0) },
            details: details
        )

        do {
            let response: HomeBillResponse = try await api.request(
                HomesEndpoints.createBill(homeId: homeId, request: request)
            )
            createdBillId = response.bill.id
            currentStep = .success
            Analytics.track(.ctaAddBillSubmit(result: .success))
        } catch {
            submitError = (error as? APIError)?.errorDescription
                ?? "Couldn't add this bill."
            Analytics.track(.ctaAddBillSubmit(result: .error))
        }
    }

    /// `yyyy-MM-dd` formatter built per-call so we don't have to manage a
    /// static `DateFormatter` under Swift 6 strict concurrency.
    private static func formatISODate(_ date: Date) -> String {
        let f = DateFormatter()
        f.locale = Locale(identifier: "en_US_POSIX")
        f.timeZone = TimeZone(secondsFromGMT: 0)
        f.dateFormat = "yyyy-MM-dd"
        return f.string(from: date)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/bills/AddBillWizardViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber")

package app.pantopus.android.ui.screens.homes.bills

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.analytics.Analytics
import app.pantopus.android.data.analytics.AnalyticsEvent
import app.pantopus.android.data.analytics.AnalyticsResult
import app.pantopus.android.data.api.models.homes.CreateBillRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.ui.screens.shared.wizard.WizardChrome
import app.pantopus.android.ui.screens.shared.wizard.WizardLeadingControl
import app.pantopus.android.ui.screens.shared.wizard.WizardModel
import app.pantopus.android.ui.screens.shared.wizard.WizardProgressLabel
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.math.BigDecimal
import java.time.LocalDate
import java.time.format.DateTimeFormatter
import javax.inject.Inject

/** Step identifiers for the Add Bill wizard. */
enum class AddBillStep { Details, Schedule, Review, Success }

/** Recurrence options for step 2. */
enum class AddBillSchedule(
    val label: String,
    val detailsKey: String,
) {
    OneTime("One-time", "one_time"),
    Monthly("Recurring monthly", "monthly"),
    Quarterly("Recurring quarterly", "quarterly"),
    Yearly("Recurring yearly", "yearly"),
}

/** Outbound events for the host to react to. */
sealed interface AddBillEvent {
    data object Dismiss : AddBillEvent

    data class Created(val billId: String) : AddBillEvent
}

/** Nav-arg key. */
const val ADD_BILL_HOME_ID_KEY = "homeId"

@HiltViewModel
class AddBillWizardViewModel
    @Inject
    constructor(
        private val repo: HomesRepository,
        savedStateHandle: SavedStateHandle,
    ) : ViewModel(), WizardModel {
        private val homeId: String = checkNotNull(savedStateHandle[ADD_BILL_HOME_ID_KEY]) {
            "AddBillWizardViewModel requires a $ADD_BILL_HOME_ID_KEY nav argument"
        }

        // Step 1
        var payee: String by mutableStateOf("")
        var amount: String by mutableStateOf("")
        var dueDate: LocalDate? by mutableStateOf(null)

        // Step 2
        var schedule: AddBillSchedule by mutableStateOf(AddBillSchedule.OneTime)

        private val _currentStep = MutableStateFlow(AddBillStep.Details)
        val currentStep: StateFlow<AddBillStep> = _currentStep.asStateFlow()

        private val _isSubmitting = MutableStateFlow(false)
        val isSubmitting: StateFlow<Boolean> = _isSubmitting.asStateFlow()

        private val _submitError = MutableStateFlow<String?>(null)
        val submitError: StateFlow<String?> = _submitError.asStateFlow()

        private val _events = MutableStateFlow<AddBillEvent?>(null)
        val events: StateFlow<AddBillEvent?> = _events.asStateFlow()

        private var createdBillId: String? = null

        override val chrome: WizardChrome
            get() =
                when (_currentStep.value) {
                    AddBillStep.Details ->
                        WizardChrome(
                            title = "Add a bill",
                            progressLabel = WizardProgressLabel.StepOf(current = 1, total = 3),
                            progressFraction = 1f / 3f,
                            leading = WizardLeadingControl.Close,
                            primaryCtaLabel = "Next",
                            primaryCtaEnabled = detailsValid(),
                            isSubmitting = false,
                            dirty = isDirty(),
                            showsProgressBar = true,
                        )
                    AddBillStep.Schedule ->
                        WizardChrome(
                            title = "Schedule",
                            progressLabel = WizardProgressLabel.StepOf(current = 2, total = 3),
                            progressFraction = 2f / 3f,
                            leading = WizardLeadingControl.Back,
                            primaryCtaLabel = "Next",
                            primaryCtaEnabled = true,
                            isSubmitting = false,
                            dirty = isDirty(),
                            showsProgressBar = true,
                        )
                    AddBillStep.Review ->
                        WizardChrome(
                            title = "Review",
                            progressLabel = WizardProgressLabel.StepOf(current = 3, total = 3),
                            progressFraction = 1f,
                            leading = WizardLeadingControl.Back,
                            primaryCtaLabel = "Add bill",
                            primaryCtaEnabled = !_isSubmitting.value,
                            isSubmitting = _isSubmitting.value,
                            dirty = isDirty(),
                            showsProgressBar = true,
                        )
                    AddBillStep.Success ->
                        WizardChrome(
                            title = "Bill added",
                            progressLabel = WizardProgressLabel.Hidden,
                            progressFraction = null,
                            leading = WizardLeadingControl.Close,
                            primaryCtaLabel = "Done",
                            primaryCtaEnabled = true,
                            isSubmitting = false,
                            dirty = false,
                            showsProgressBar = false,
                        )
                }

        override fun onLeading() {
            when (_currentStep.value) {
                AddBillStep.Details -> _events.value = AddBillEvent.Dismiss
                AddBillStep.Schedule -> _currentStep.value = AddBillStep.Details
                AddBillStep.Review -> _currentStep.value = AddBillStep.Schedule
                AddBillStep.Success -> _events.value = AddBillEvent.Dismiss
            }
        }

        override fun onDiscard() {
            _events.value = AddBillEvent.Dismiss
        }

        override fun onPrimary() {
            when (_currentStep.value) {
                AddBillStep.Details -> {
                    _currentStep.value = AddBillStep.Schedule
                    Analytics.track(AnalyticsEvent.ScreenAddBillWizardStepViewed(2, "schedule"))
                }
                AddBillStep.Schedule -> {
                    _currentStep.value = AddBillStep.Review
                    Analytics.track(AnalyticsEvent.ScreenAddBillWizardStepViewed(3, "review"))
                }
                AddBillStep.Review -> submit()
                AddBillStep.Success -> {
                    val id = createdBillId
                    _events.value = id?.let(AddBillEvent::Created) ?: AddBillEvent.Dismiss
                }
            }
        }

        fun consumeEvent() {
            _events.value = null
        }

        fun parsedAmount(): BigDecimal? {
            val trimmed = amount.trim()
            if (trimmed.isEmpty()) return null
            return trimmed.toBigDecimalOrNull()?.takeIf { it > BigDecimal.ZERO }
        }

        fun detailsValid(): Boolean = payee.trim().isNotEmpty() && parsedAmount() != null

        fun isDirty(): Boolean =
            payee.trim().isNotEmpty() ||
                amount.trim().isNotEmpty() ||
                dueDate != null ||
                schedule != AddBillSchedule.OneTime

        private fun submit() {
            val amountValue = parsedAmount() ?: return
            if (_isSubmitting.value) return
            _isSubmitting.value = true
            _submitError.value = null
            viewModelScope.launch {
                val details = buildMap {
                    put("schedule", schedule.detailsKey)
                    if (schedule != AddBillSchedule.OneTime) {
                        put("frequency", schedule.detailsKey)
                    }
                }
                val request =
                    CreateBillRequest(
                        billType = "other",
                        providerName = payee.trim(),
                        amount = amountValue,
                        dueDate = dueDate?.format(DateTimeFormatter.ISO_LOCAL_DATE),
                        details = details,
                    )
                when (val result = repo.createHomeBill(homeId, request)) {
                    is NetworkResult.Success -> {
                        createdBillId = result.data.bill.id
                        _isSubmitting.value = false
                        _currentStep.value = AddBillStep.Success
                        Analytics.track(AnalyticsEvent.CtaAddBillSubmit(AnalyticsResult.SUCCESS))
                    }
                    is NetworkResult.Failure -> {
                        _isSubmitting.value = false
                        _submitError.value = result.error.message
                        Analytics.track(AnalyticsEvent.CtaAddBillSubmit(AnalyticsResult.ERROR))
                    }
                }
            }
        }
    }
